A3.md

```diff
@@ -45,40 +45,42 @@ One downside to this approach is the potential for the tracer to bloat memory. F
 The implementation will be done in the C-core first, with Java and Go to follow. The tracer object will hold a linked list of trace node and a linked list of subchannel tracers. Each of the subchannel tracers will hold their own linked lists of trace nodes.
 
 ```
+// One node of tracing data
 struct grpc_trace_node {
   grpc_slice* data;
   grpc_error* error;
   gpr_timespec time;
   grpc_connectivity_state connectivity_state;
   grpc_trace_node* next;
+
+  // this last field is used for efficient garbage collection of the
+  // subchannels. Once the node that traces a subchannel going inactive is
+  // overwritten, the tracing object for that subchannel is freed.
+  grpc_subchannel_tracer* free_this_subchannel;
 };
 
 struct trace_node_list {
   int size;
   grpc_trace_node* head_trace;
   grpc_trace_node* tail_trace;
-  grpc_subchannel_tracer* referrenced_subchannel;
 };
 
 /* the channel tracing object */
 struct grpc_channel_tracer {
   gpr_mu tracer_mu; 
   uint32_t num_nodes_logged;
   trace_node_list node_list;
-  grpc_subchannel_tracer* head_subchannel;
-  grpc_subchannel_tracer* tail_subchannel;
-};
-
-/* the subchannel tracing object */
-struct grpc_subchannel_tracer {
-  gpr_mu tracer_mu;
-  gpr_refcount refs;
-  trace_node_list node_list;
   uint32_t num_subchannels_logged;
-  grpc_subchannel_tracer* next;
-  grpc_subchannel_tracer* prev;
+  gpr_timespec time_created;
+
+  // For the parent channel, these fields are the head and tail of the linked
+  // list of subchannels. For the subchannels, these fields are the next and
+  // prev pointers.
+  grpc_subchannel_tracer* head_prev_subchannel;
```

Comment:
If these fields do stick around, they probably need be of type grpc_channel_tracer, not grpc_subchannel_tracer.